docs/proposals/201912_thanos_binary_index_header.md

```diff
@@ -0,0 +1,215 @@
+scans through `posting` section of the TSDB index. Having to fetch whole postings section just to get size of each posting 
+makes this proposal less valuable as we still need to download big part of index and traverse through it instead of what we propose in [#Proposal](./201912_thanos_binary_index_header.md#proposal)
+
+For series we don't know the exact size either, however we estimate max size of each series to be 64*1024. It depends on sane number of label pairs and chunks per series.
+We have really only one potential case when this was too low: https://github.com/thanos-io/thanos/issues/552. Decision about series this was made here: https://github.com/thanos-io/thanos/issues/146
+
+For postings it's more tricky as it depends on number of series in which given label pair exists. For worst case it can be even million label-pair for popular pairs like `__name__=http_request_http_duration_bucket` etc.
+
+We have few options:
+
+* Encode this posting size in TSDB index `PostingOffset`. 
+* Scan postings to fetch, which is something we wanted to avoid when building `index-header` without downloading full TSDB index.
+* Estimate some large value (will overfetch)
+* Estimate casual value and retry fetching remaining size on demand for data consistency.
+
+### Unexpected memory usage
+
+Users care the most about surprising spikes in memory usage. Currently the Store Gateway caches the whole index-cache.json. While it's silly to do so for all blocks, this 
+will happen anyway if query will span over large number of blocks and series. This means that while baseline memory will be reduced, baseline vs request memory difference will be even more noticeable.
+
+This tradeoff is acceptable, due to total memory used for all operation should be much smaller. Additionally such query spanning all of the blocks and series are unlikely and should be blocked by simple `sample` limit.
+
+### Benchmark
+
+How to micro-benchmark such change? mmaping is outside of Go run time, which counts allocations etc.
+ 
+### Do we need to mmap?
+
+mmap adds a lot of complexity and confusion especially around monitoring it's memory usage as it does not appear on Go profiles.
+
+While mmap is great for random access against a big file, in fact the current implementation of the [FileReader](https://github.com/prometheus/prometheus/blob/de0a772b8e7d27dc744810a1a693d97be027049a/tsdb/index/index.go#L664)
+reallocates symbols, offsets and label name=value pairs while reading. This really defies the purpose of mmap as we want to combine all info in a dense few sequential sections of index-header binary format, 
+this file will be mostly read sequentially. Still, label values can be accessed randomly, that's why we propose to start with mmap straight away. 
+
+### Is initial long startup, really a problem?
+
+After initial startup with persistent disk, next startup should be quick due to cached files on disk for old blocks.
```

Comment:
> After initial startup with persistent disk, next startup should be quick due to cached files on disk for old blocks.

Another use case is scaling. When you scale up your cluster because of high load, you want the store to be ready as quickly as possible. The initial long startup delays it.

Reply:
yup, adding